causalnex/structure/dynotears.py

```diff
@@ -0,0 +1,745 @@
+    return "{var}_lag{l_val}".format(var=idx_col[int(idx)], l_val=lag_val)
+
+
+def _cut_dataframes_on_discontinuity_points(
+    time_series: List[pd.DataFrame],
+) -> List[np.ndarray]:
+    """
+    Helper function for `from_pandas_dynamic`
+    Receive a list of dataframes. For each dataframe, cut the points of discontinuity as two different dataframes.
+    Discontinuities are determined by the indexes.
+
+    For Example:
+    If the following is a dataframe:
+        index   variable_1  variable_2
+        1       X           X
+        2       X           X
+        3       X           X
+        4       X           X
+        8       X           X               <- discontinuity point
+        9       X           X
+        10      X           X
+
+    We cut this dataset in two:
+
+        index   variable_1  variable_2
+        1       X           X
+        2       X           X
+        3       X           X
+        4       X           X
+
+        and:
+        index   variable_1  variable_2
+        8       X           X
+        9       X           X
+        10      X           X
+
+
+    Args:
+        time_series: list of dataframes representing various realisations of a same time series
+
+    Returns:
+        List of np.ndarrays representing the pieces of the input datasets with no discontinuity
+
+    """
+    time_series_realisations = []
+    for t in time_series:
+        cutting_points = np.where(np.diff(t.index) > 1)[0]
+        cutting_points = [0] + list(cutting_points + 1) + [len(t)]
+        for start, end in zip(cutting_points[:-1], cutting_points[1:]):
+            time_series_realisations.append(t.iloc[start:end, :].values)
+    return time_series_realisations
+
+
+def _convert_realisations_into_dynotears_format(
+    realisations: List[np.ndarray], p: int
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Given a list of realisations of a time series, convert it to the format received by the dynotears algorithm.
+    Each realisation on `realisations` is a realisation of the time series, where the time dimension is represented by
+    the rows.
+        - The higher the row, the higher the time index
+        - The data is complete, meaning that the difference between two time stamps is equal one
+    Args:
+        realisations: a list of realisations of a time series
+        p: the number of lagged columns to create
+
+    Returns:
+        X and Y as in the SVAR model and DYNOTEARS paper. I.e. X being representing X(m,t) and Y the concatenated
+        differences [X(m,t-1) | X(m,t-2) | ... | X(m,t-p)]
+    """
+    X = np.concatenate([realisation[p:] for realisation in realisations], axis=0)
+    y_lag_list = [
+        np.concatenate([realisation[p - i - 1 : -i - 1] for i in range(p)], axis=1)
+        for realisation in realisations
+    ]
+    y_lag = np.concatenate(y_lag_list, axis=0)
+
+    return X, y_lag
+
+
+class DynamicDataTransformer(TransformerMixin):
```

Comment:
Would it make sense to put this in a separate transformers.py file so that the pytorch implementation can make use of it down the line?

Reply:
also might want to also inherit from `BaseEstimator` to allow the transformer to be used in a pipeline:

```suggestion
class DynamicDataTransformer(TransformerMixin, BaseEstimator):
```